pkg/worker/task_runner.go

```diff
@@ -54,101 +54,128 @@ func NewTaskRunnerWithApiClient(
 
-func (c *TaskRunner) startWorker(taskType string, executeFunction model.TaskExecuteFunction, threadCount int, pollIntervalInMillis int, taskDomain string) {
+func (c *TaskRunner) startWorker(taskType string, executeFunction model.TaskExecuteFunction, threadCount int, pollInterval time.Duration, taskDomain string) error {
 	var domain optional.String
 	if taskDomain != "" {
 		domain = optional.NewString(taskDomain)
 	}
-	previousMaxAllowedWorkers := c.getMaxAllowedWorkers(taskType)
-	c.increaseMaxAllowedWorkers(taskType, threadCount)
+	previousMaxAllowedWorkers, err := c.getMaxAllowedWorkers(taskType)
+	if err != nil {
+		return err
+	}
+	err = c.increaseMaxAllowedWorkers(taskType, threadCount)
+	if err != nil {
+		return err
+	}
 	if previousMaxAllowedWorkers == 0 {
 		c.workerWaitGroup.Add(1)
-		go c.pollAndExecute(taskType, executeFunction, pollIntervalInMillis, domain)
+		go c.pollAndExecute(taskType, executeFunction, pollInterval, domain)
 	}
-	log.Info(
+	logrus.Debug(
 		"Started worker for task: ", taskType,
 		", threadCount / batchSize: ", threadCount,
-		", polling interval: ", pollIntervalInMillis, "ms",
+		", polling interval: ", pollInterval.Milliseconds(), "ms",
 	)
+	return nil
 }
 
-func (c *TaskRunner) pollAndExecute(taskType string, executeFunction model.TaskExecuteFunction, pollingInterval int, domain optional.String) {
-	defer func() {
-		c.workerWaitGroup.Done()
-		concurrency.OnError("poll_and_execute")
-		log.Warning(
-			"Panic at pollAndExecute",
-			", taskType: ", taskType,
-			", pollingInterval: ", pollingInterval,
-			", domain: ", domain.Value(),
-		)
-	}()
+func (c *TaskRunner) pollAndExecute(taskType string, executeFunction model.TaskExecuteFunction, pollInterval time.Duration, domain optional.String) error {
+	defer concurrency.HandlePanicError("poll_and_execute")
 	for c.isWorkerAlive(taskType) {
-		c.runBatch(taskType, executeFunction, pollingInterval, domain)
+		isTaskQueueEmpty, err := c.runBatch(taskType, executeFunction, pollInterval, domain)
+		if err != nil {
+			logrus.Warning(
+				"Failed to poll and execute",
+				", reason: ", err.Error(),
+				", taskType: ", taskType,
+				", pollInterval: ", pollInterval.Milliseconds(), "ms",
+				", domain: ", domain,
+			)
+		} else if isTaskQueueEmpty {
+			logrus.Debug("No tasks available for: ", taskType)
+			time.Sleep(pollInterval)
+		}
 	}
+	c.workerWaitGroup.Done()
+	return nil
 }
 
-func (c *TaskRunner) runBatch(taskType string, executeFunction model.TaskExecuteFunction, pollingInterval int, domain optional.String) {
-	batchSize := c.getAvailableWorkerAmount(taskType)
+func (c *TaskRunner) runBatch(taskType string, executeFunction model.TaskExecuteFunction, pollInterval time.Duration, domain optional.String) (bool, error) {
+	batchSize, err := c.getAvailableWorkerAmount(taskType)
+	if err != nil {
+		return false, err
+	}
 	if batchSize < 1 {
-		sleep(1)
-		return
+		// TODO wait until there is available workers
+		time.Sleep(10 * time.Millisecond)
```

Comment:
10 ms is too much, let's keep it to 1 ms

Reply:
done